aten/src/ATen/native/mps/operations/Quantized.mm

```diff
@@ -22,41 +22,89 @@
+  using type = bfloat4;
+};
+#endif
 
 template<typename T, unsigned groupSize>
 kernel void int4pack_mm(
     constant T                 * A              [[buffer(0)]],
     constant uchar             * B              [[buffer(1)]],
     constant T                 * scalesAndZeros [[buffer(2)]],
     device   T                 * outputData     [[buffer(3)]],
-    constant uint3             & sizes          [[buffer(4)]],
+    constant uint3             & sizes          [[buffer(4)]], // M, K, N
     uint2                        thread_index   [[thread_position_in_grid]]) {
-    const uint lda = sizes.y;
-    const uint ldc = sizes.z;
-    const uint m = thread_index.y; // 0..sizes.x-1
-    const uint n = thread_index.x; // 0..sizes.z-1
-    const uint nb = n / 32;
-    const uint ldb = min(32U,  sizes.z - nb * 32);
-    const uint32_t k_block = (sizes.y + groupSize - 1) / groupSize;
-    constant T *A_ptr = A + m * lda;
-    constant uchar *B_ptr = B + (nb * 16 * sizes.y);
-
-    float rc = 0.0;
+    const uint K = sizes.y;
+    const uint N = sizes.z;
+    const uint m = thread_index.y; // 0..M/4-1
+    const uint n = thread_index.x; // 0..N/4-1
+    const uint nb = n / 8;
+    const uint ldb = min(32U,  N - nb * 32);
+    const uint32_t k_block = (K + groupSize - 1) / groupSize;
+
+    using vecT = typename Vec4Type<T>::type;
+    constant vecT *A_ptr = reinterpret_cast<constant vecT *>(A + m * 4 * K);
+    //constant uchar2 *B_ptr = reinterpret_cast<constant uchar2 *>(B + (nb * 16 * K));
+    constant uchar *B_ptr = B + (nb * 16 * K);
+
+    float4x4 rc;
+    for(int j = 0; j < 4; ++j) {
+      rc[j] = float4(0.0);
+    }
     uint k = 0;
     for (uint32_t kb = 0; kb < k_block ; kb ++) {
-      const T scale = scalesAndZeros[(kb * ldc + n) * 2 + 0];
-      const T zero = scalesAndZeros[(kb * ldc + n) * 2 + 1] - scale * T(8);
-      for(uint idx = 0; idx < groupSize && k < sizes.y; idx++, k++) {
-        const auto a_val = float(A_ptr[k]);
-        uchar b_val = B_ptr[(k * ldb + (n % 32))/2];
-        b_val = (n & 1) == 0 ? b_val & 0x0f : (b_val >> 4);
-        rc += a_val * float(scale * T(b_val) + zero);
+      const T scale0 = scalesAndZeros[(kb * N + 4 * n) * 2 + 0];
+      const T zero0 = scalesAndZeros[(kb * N + 4 * n) * 2 + 1] - scale0 * T(8);
+
+      const T scale1 = scalesAndZeros[(kb * N + 4 * n + 1) * 2 + 0];
+      const T zero1 = scalesAndZeros[(kb * N + 4 * n + 1) * 2 + 1] - scale1 * T(8);
+
+      const T scale2 = scalesAndZeros[(kb * N + 4 * n + 2) * 2 + 0];
+      const T zero2 = scalesAndZeros[(kb * N + 4 * n + 2) * 2 + 1] - scale2 * T(8);
+
+      const T scale3 = scalesAndZeros[(kb * N + 4 * n + 3) * 2 + 0];
+      const T zero3 = scalesAndZeros[(kb * N + 4 * n + 3) * 2 + 1] - scale3 * T(8);
+
+      const float4 scales = float4(scale0, scale1, scale2, scale3);
+      const float4 zeros = float4(zero0, zero1, zero2, zero3);
+
+      for(uint idx = 0; idx < groupSize && k < K; idx += 4, k += 4) {
```

Comment:
what if `K % 4 != 0`? I don't see a check for that

Reply:
K is supposed to be a multiple of 4. Actually, it should be a multiple of 32, based on the TORCH_CHECKS here:
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LinearAlgebra.cpp?fbclid=IwAR3k1YtoBlM8b0H2AkMfdPr10Yr4nIhHHVFgXEp1o9-TmeSDk39ZM58lrUw#L3439